// @compodoc/compodoc/dist/index-cli.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var application = require('./application-54cd2170.js');
var fs = require('fs-extra');
var path = require('path');
var Ast = require('ts-simple-ast');
var Ast__default = _interopDefault(Ast);
require('live-server');
require('lodash');
require('i18next');
require('handlebars');
require('semver');
require('json5');

var glob = require('glob');
var ParserUtil = /** @class */ (function () {
    function ParserUtil() {
        this._globFiles = [];
    }
    ParserUtil.prototype.init = function (exclude, cwd) {
        this._files = exclude;
        this._cwd = cwd;
        var i = 0;
        var len = exclude.length;
        for (i; i < len; i++) {
            this._globFiles = this._globFiles.concat(glob.sync(exclude[i], { cwd: this._cwd }));
        }
    };
    ParserUtil.prototype.testFilesWithCwdDepth = function () {
        var i = 0;
        var len = this._files.length;
        var result = {
            status: true,
            level: 0
        };
        for (i; i < len; i++) {
            var elementPath = path.resolve(this._cwd + path.sep, this._files[i]);
            if (elementPath.indexOf(this._cwd) === -1) {
                result.status = false;
                var level = this._files[i].match(/\..\//g).length;
                if (level > result.level) {
                    result.level = level;
                }
            }
        }
        return result;
    };
    ParserUtil.prototype.updateCwd = function (cwd, level) {
        var _cwd = cwd, _rewind = '';
        for (var i = 0; i < level; i++) {
            _rewind += '../';
        }
        _cwd = path.resolve(_cwd, _rewind);
        return _cwd;
    };
    ParserUtil.prototype.testFile = function (file) {
        var _this = this;
        var i = 0;
        var len = this._files.length;
        var fileBasename = path.basename(file);
        var fileNameInCwd = file.replace(this._cwd + path.sep, '');
        var result = false;
        if (path.sep === '\\') {
            fileNameInCwd = fileNameInCwd.replace(new RegExp('\\' + path.sep, 'g'), '/');
        }
        for (i; i < len; i++) {
            if (glob.hasMagic(this._files[i]) && this._globFiles.length > 0) {
                var resultGlobSearch = this._globFiles.findIndex(function (element) {
                    var elementPath = path.resolve(_this._cwd + path.sep, element);
                    var elementPathInCwd = elementPath.replace(_this._cwd + path.sep, '');
                    elementPathInCwd = elementPathInCwd.replace(new RegExp('\\' + path.sep, 'g'), '/');
                    return elementPathInCwd === fileNameInCwd;
                });
                result = resultGlobSearch !== -1;
            }
            else {
                result = fileNameInCwd === this._files[i];
            }
            if (result) {
                break;
            }
        }
        return result;
    };
    return ParserUtil;
}());

var cosmiconfig = require('cosmiconfig');
var os = require('os');
var osName = require('os-name');
var pkg = require('../package.json');
var program = require('commander');
var cosmiconfigModuleName = 'compodoc';
var scannedFiles = [];
var excludeFiles;
var includeFiles;
var cwd = process.cwd();
process.setMaxListeners(0);
var CliApplication = /** @class */ (function (_super) {
    application.__extends(CliApplication, _super);
    function CliApplication() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Run compodoc from the command line.
     */
    CliApplication.prototype.start = function () {
        var _this = this;
        function list(val) {
            return val.split(',');
        }
        program
            .version(pkg.version)
            .usage('<src> [options]')
            .option('-c, --config [config]', 'A configuration file : .compodocrc, .compodocrc.json, .compodocrc.yaml or compodoc property in package.json')
            .option('-p, --tsconfig [config]', 'A tsconfig.json file')
            .option('-d, --output [folder]', 'Where to store the generated documentation', application.COMPODOC_DEFAULTS.folder)
            .option('-y, --extTheme [file]', 'External styling theme file')
            .option('-n, --name [name]', 'Title documentation', application.COMPODOC_DEFAULTS.title)
            .option('-a, --assetsFolder [folder]', 'External assets folder to copy in generated documentation folder')
            .option('-o, --open [value]', 'Open the generated documentation')
            .option('-t, --silent', "In silent mode, log messages aren't logged in the console", false)
            .option('-s, --serve', 'Serve generated documentation (default http://localhost:8080/)', false)
            .option('--host [host]', 'Change default host address')
            .option('-r, --port [port]', 'Change default serving port', application.COMPODOC_DEFAULTS.port)
            .option('-w, --watch', 'Watch source files after serve and force documentation rebuild', false)
            .option('-e, --exportFormat [format]', 'Export in specified format (json, html)', application.COMPODOC_DEFAULTS.exportFormat)
            .option('--files [files]', 'Files provided by external tool, used for coverage test')
            .option('--language [language]', 'Language used for the generated documentation (de-DE, en-US, es-ES, fr-FR, hu-HU, it-IT, ja-JP, nl-NL, pt-BR, sk-SK, zh-CN)', application.COMPODOC_DEFAULTS.language)
            .option('--theme [theme]', "Choose one of available themes, default is 'gitbook' (laravel, original, material, postmark, readthedocs, stripe, vagrant)")
            .option('--hideGenerator', 'Do not print the Compodoc link at the bottom of the page', false)
            .option('--toggleMenuItems <items>', "Close by default items in the menu values : ['all'] or one of these ['modules','components','directives','controllers','classes','injectables','guards','interfaces','interceptors','pipes','miscellaneous','additionalPages']", list, application.COMPODOC_DEFAULTS.toggleMenuItems)
            .option('--navTabConfig <tab configs>', "List navigation tab objects in the desired order with two string properties (\"id\" and \"label\"). Double-quotes must be escaped with '\\'. Available tab IDs are \"info\", \"readme\", \"source\", \"templateData\", \"styleData\", \"tree\", and \"example\". Note: Certain tabs will only be shown if applicable to a given dependency", list, JSON.stringify(application.COMPODOC_DEFAULTS.navTabConfig))
            .option('--templates [folder]', 'Path to directory of Handlebars templates to override built-in templates')
            .option('--includes [path]', 'Path of external markdown files to include')
            .option('--includesName [name]', 'Name of item menu of externals markdown files', application.COMPODOC_DEFAULTS.additionalEntryName)
            .option('--coverageTest [threshold]', 'Test command of documentation coverage with a threshold (default 70)')
            .option('--coverageMinimumPerFile [minimum]', 'Test command of documentation coverage per file with a minimum (default 0)')
            .option('--coverageTestThresholdFail [true|false]', 'Test command of documentation coverage (global or per file) will fail with error or just warn user (true: error, false: warn)', application.COMPODOC_DEFAULTS.coverageTestThresholdFail)
            .option('--coverageTestShowOnlyFailed', 'Display only failed files for a coverage test')
            .option('--unitTestCoverage [json-summary]', 'To include unit test coverage, specify istanbul JSON coverage summary file')
            .option('--disableSourceCode', 'Do not add source code tab and links to source code', false)
            .option('--disableDomTree', 'Do not add dom tree tab', false)
            .option('--disableTemplateTab', 'Do not add template tab', false)
            .option('--disableStyleTab', 'Do not add style tab', false)
            .option('--disableGraph', 'Do not add the dependency graph', false)
            .option('--disableCoverage', 'Do not add the documentation coverage report', false)
            .option('--disablePrivate', 'Do not show private in generated documentation', false)
            .option('--disableProtected', 'Do not show protected in generated documentation', false)
            .option('--disableInternal', 'Do not show @internal in generated documentation', false)
            .option('--disableLifeCycleHooks', 'Do not show Angular lifecycle hooks in generated documentation', false)
            .option('--disableRoutesGraph', 'Do not add the routes graph', application.COMPODOC_DEFAULTS.disableRoutesGraph)
            .option('--disableSearch', 'Do not add the search input', false)
            .option('--disableDependencies', 'Do not add the dependencies list', application.COMPODOC_DEFAULTS.disableDependencies)
            .option('--minimal', 'Minimal mode with only documentation. No search, no graph, no coverage.', false)
            .option('--customFavicon [path]', 'Use a custom favicon')
            .option('--customLogo [path]', 'Use a custom logo')
            .option('--gaID [id]', 'Google Analytics tracking ID')
            .option('--gaSite [site]', 'Google Analytics site name', application.COMPODOC_DEFAULTS.gaSite)
            .option('--maxSearchResults [maxSearchResults]', 'Max search results on the results page. To show all results, set to 0', application.COMPODOC_DEFAULTS.maxSearchResults)
            .parse(process.argv);
        var outputHelp = function () {
            program.outputHelp();
            process.exit(1);
        };
        var configExplorer = cosmiconfig(cosmiconfigModuleName);
        var configExplorerResult;
        var configFile = {};
        if (program.config) {
            var configFilePath = program.config;
            var testConfigFilePath = configFilePath.match(process.cwd());
            if (testConfigFilePath && testConfigFilePath.length > 0) {
                configFilePath = configFilePath.replace(process.cwd() + path.sep, '');
            }
            configExplorerResult = configExplorer.loadSync(path.resolve(configFilePath));
        }
        else {
            configExplorerResult = configExplorer.searchSync();
        }
        if (configExplorerResult) {
            if (typeof configExplorerResult.config !== 'undefined') {
                configFile = configExplorerResult.config;
            }
        }
        if (configFile.output) {
            application.Configuration.mainData.output = configFile.output;
        }
        if (program.output && program.output !== application.COMPODOC_DEFAULTS.folder) {
            application.Configuration.mainData.output = program.output;
        }
        if (configFile.extTheme) {
            application.Configuration.mainData.extTheme = configFile.extTheme;
        }
        if (program.extTheme) {
            application.Configuration.mainData.extTheme = program.extTheme;
        }
        if (configFile.language) {
            application.Configuration.mainData.language = configFile.language;
        }
        if (program.language) {
            application.Configuration.mainData.language = program.language;
        }
        if (configFile.theme) {
            application.Configuration.mainData.theme = configFile.theme;
        }
        if (program.theme) {
            application.Configuration.mainData.theme = program.theme;
        }
        if (configFile.name) {
            application.Configuration.mainData.documentationMainName = configFile.name;
        }
        if (program.name && program.name !== application.COMPODOC_DEFAULTS.title) {
            application.Configuration.mainData.documentationMainName = program.name;
        }
        if (configFile.assetsFolder) {
            application.Configuration.mainData.assetsFolder = configFile.assetsFolder;
        }
        if (program.assetsFolder) {
            application.Configuration.mainData.assetsFolder = program.assetsFolder;
        }
        if (configFile.open) {
            application.Configuration.mainData.open = configFile.open;
        }
        if (program.open) {
            application.Configuration.mainData.open = program.open;
        }
        if (configFile.toggleMenuItems) {
            application.Configuration.mainData.toggleMenuItems = configFile.toggleMenuItems;
        }
        if (program.toggleMenuItems &&
            program.toggleMenuItems !== application.COMPODOC_DEFAULTS.toggleMenuItems) {
            application.Configuration.mainData.toggleMenuItems = program.toggleMenuItems;
        }
        if (configFile.templates) {
            application.Configuration.mainData.templates = configFile.templates;
        }
        if (program.templates) {
            application.Configuration.mainData.templates = program.templates;
        }
        if (configFile.navTabConfig) {
            application.Configuration.mainData.navTabConfig = configFile.navTabConfig;
        }
        if (program.navTabConfig &&
            JSON.parse(program.navTabConfig).length !== application.COMPODOC_DEFAULTS.navTabConfig.length) {
            application.Configuration.mainData.navTabConfig = JSON.parse(program.navTabConfig);
        }
        if (configFile.includes) {
            application.Configuration.mainData.includes = configFile.includes;
        }
        if (program.includes) {
            application.Configuration.mainData.includes = program.includes;
        }
        if (configFile.includesName) {
            application.Configuration.mainData.includesName = configFile.includesName;
        }
        if (program.includesName &&
            program.includesName !== application.COMPODOC_DEFAULTS.additionalEntryName) {
            application.Configuration.mainData.includesName = program.includesName;
        }
        if (configFile.silent) {
            application.logger.silent = false;
        }
        if (program.silent) {
            application.logger.silent = false;
        }
        if (configFile.serve) {
            application.Configuration.mainData.serve = configFile.serve;
        }
        if (program.serve) {
            application.Configuration.mainData.serve = program.serve;
        }
        if (configFile.host) {
            application.Configuration.mainData.host = configFile.host;
            application.Configuration.mainData.hostname = configFile.host;
        }
        if (program.host) {
            application.Configuration.mainData.host = program.host;
            application.Configuration.mainData.hostname = program.host;
        }
        if (configFile.port) {
            application.Configuration.mainData.port = configFile.port;
        }
        if (program.port && program.port !== application.COMPODOC_DEFAULTS.port) {
            application.Configuration.mainData.port = program.port;
        }
        if (configFile.watch) {
            application.Configuration.mainData.watch = configFile.watch;
        }
        if (program.watch) {
            application.Configuration.mainData.watch = program.watch;
        }
        if (configFile.exportFormat) {
            application.Configuration.mainData.exportFormat = configFile.exportFormat;
        }
        if (program.exportFormat && program.exportFormat !== application.COMPODOC_DEFAULTS.exportFormat) {
            application.Configuration.mainData.exportFormat = program.exportFormat;
        }
        if (configFile.hideGenerator) {
            application.Configuration.mainData.hideGenerator = configFile.hideGenerator;
        }
        if (program.hideGenerator) {
            application.Configuration.mainData.hideGenerator = program.hideGenerator;
        }
        if (configFile.coverageTest) {
            application.Configuration.mainData.coverageTest = true;
            application.Configuration.mainData.coverageTestThreshold =
                typeof configFile.coverageTest === 'string'
                    ? parseInt(configFile.coverageTest, 10)
                    : application.COMPODOC_DEFAULTS.defaultCoverageThreshold;
        }
        if (program.coverageTest) {
            application.Configuration.mainData.coverageTest = true;
            application.Configuration.mainData.coverageTestThreshold =
                typeof program.coverageTest === 'string'
                    ? parseInt(program.coverageTest, 10)
                    : application.COMPODOC_DEFAULTS.defaultCoverageThreshold;
        }
        if (configFile.coverageMinimumPerFile) {
            application.Configuration.mainData.coverageTestPerFile = true;
            application.Configuration.mainData.coverageMinimumPerFile =
                typeof configFile.coverageMinimumPerFile === 'string'
                    ? parseInt(configFile.coverageMinimumPerFile, 10)
                    : application.COMPODOC_DEFAULTS.defaultCoverageMinimumPerFile;
        }
        if (program.coverageMinimumPerFile) {
            application.Configuration.mainData.coverageTestPerFile = true;
            application.Configuration.mainData.coverageMinimumPerFile =
                typeof program.coverageMinimumPerFile === 'string'
                    ? parseInt(program.coverageMinimumPerFile, 10)
                    : application.COMPODOC_DEFAULTS.defaultCoverageMinimumPerFile;
        }
        if (configFile.coverageTestThresholdFail) {
            application.Configuration.mainData.coverageTestThresholdFail =
                configFile.coverageTestThresholdFail === 'false' ? false : true;
        }
        if (program.coverageTestThresholdFail) {
            application.Configuration.mainData.coverageTestThresholdFail =
                program.coverageTestThresholdFail === 'false' ? false : true;
        }
        if (configFile.coverageTestShowOnlyFailed) {
            application.Configuration.mainData.coverageTestShowOnlyFailed =
                configFile.coverageTestShowOnlyFailed;
        }
        if (program.coverageTestShowOnlyFailed) {
            application.Configuration.mainData.coverageTestShowOnlyFailed = program.coverageTestShowOnlyFailed;
        }
        if (configFile.unitTestCoverage) {
            application.Configuration.mainData.unitTestCoverage = configFile.unitTestCoverage;
        }
        if (program.unitTestCoverage) {
            application.Configuration.mainData.unitTestCoverage = program.unitTestCoverage;
        }
        if (configFile.disableSourceCode) {
            application.Configuration.mainData.disableSourceCode = configFile.disableSourceCode;
        }
        if (program.disableSourceCode) {
            application.Configuration.mainData.disableSourceCode = program.disableSourceCode;
        }
        if (configFile.disableDomTree) {
            application.Configuration.mainData.disableDomTree = configFile.disableDomTree;
        }
        if (program.disableDomTree) {
            application.Configuration.mainData.disableDomTree = program.disableDomTree;
        }
        if (configFile.disableTemplateTab) {
            application.Configuration.mainData.disableTemplateTab = configFile.disableTemplateTab;
        }
        if (program.disableTemplateTab) {
            application.Configuration.mainData.disableTemplateTab = program.disableTemplateTab;
        }
        if (configFile.disableStyleTab) {
            application.Configuration.mainData.disableStyleTab = configFile.disableStyleTab;
        }
        if (program.disableStyleTab) {
            application.Configuration.mainData.disableStyleTab = program.disableStyleTab;
        }
        if (configFile.disableGraph) {
            application.Configuration.mainData.disableGraph = configFile.disableGraph;
        }
        if (program.disableGraph) {
            application.Configuration.mainData.disableGraph = program.disableGraph;
        }
        if (configFile.disableCoverage) {
            application.Configuration.mainData.disableCoverage = configFile.disableCoverage;
        }
        if (program.disableCoverage) {
            application.Configuration.mainData.disableCoverage = program.disableCoverage;
        }
        if (configFile.disablePrivate) {
            application.Configuration.mainData.disablePrivate = configFile.disablePrivate;
        }
        if (program.disablePrivate) {
            application.Configuration.mainData.disablePrivate = program.disablePrivate;
        }
        if (configFile.disableProtected) {
            application.Configuration.mainData.disableProtected = configFile.disableProtected;
        }
        if (program.disableProtected) {
            application.Configuration.mainData.disableProtected = program.disableProtected;
        }
        if (configFile.disableInternal) {
            application.Configuration.mainData.disableInternal = configFile.disableInternal;
        }
        if (program.disableInternal) {
            application.Configuration.mainData.disableInternal = program.disableInternal;
        }
        if (configFile.disableLifeCycleHooks) {
            application.Configuration.mainData.disableLifeCycleHooks = configFile.disableLifeCycleHooks;
        }
        if (program.disableLifeCycleHooks) {
            application.Configuration.mainData.disableLifeCycleHooks = program.disableLifeCycleHooks;
        }
        if (configFile.disableRoutesGraph) {
            application.Configuration.mainData.disableRoutesGraph = configFile.disableRoutesGraph;
        }
        if (program.disableRoutesGraph) {
            application.Configuration.mainData.disableRoutesGraph = program.disableRoutesGraph;
        }
        if (configFile.disableSearch) {
            application.Configuration.mainData.disableSearch = configFile.disableSearch;
        }
        if (program.disableSearch) {
            application.Configuration.mainData.disableSearch = program.disableSearch;
        }
        if (configFile.disableDependencies) {
            application.Configuration.mainData.disableDependencies = configFile.disableDependencies;
        }
        if (program.disableDependencies) {
            application.Configuration.mainData.disableDependencies = program.disableDependencies;
        }
        if (configFile.minimal) {
            application.Configuration.mainData.disableSearch = true;
            application.Configuration.mainData.disableRoutesGraph = true;
            application.Configuration.mainData.disableGraph = true;
            application.Configuration.mainData.disableCoverage = true;
        }
        if (program.minimal) {
            application.Configuration.mainData.disableSearch = true;
            application.Configuration.mainData.disableRoutesGraph = true;
            application.Configuration.mainData.disableGraph = true;
            application.Configuration.mainData.disableCoverage = true;
        }
        if (configFile.customFavicon) {
            application.Configuration.mainData.customFavicon = configFile.customFavicon;
        }
        if (program.customFavicon) {
            application.Configuration.mainData.customFavicon = program.customFavicon;
        }
        if (configFile.customLogo) {
            application.Configuration.mainData.customLogo = configFile.customLogo;
        }
        if (program.customLogo) {
            application.Configuration.mainData.customLogo = program.customLogo;
        }
        if (configFile.gaID) {
            application.Configuration.mainData.gaID = configFile.gaID;
        }
        if (program.gaID) {
            application.Configuration.mainData.gaID = program.gaID;
        }
        if (configFile.gaSite) {
            application.Configuration.mainData.gaSite = configFile.gaSite;
        }
        if (program.gaSite && program.gaSite !== application.COMPODOC_DEFAULTS.gaSite) {
            application.Configuration.mainData.gaSite = program.gaSite;
        }
        if (!this.isWatching) {
            if (!application.logger.silent) {
                console.log("Compodoc v" + pkg.version);
            }
            else {
                console.log(fs.readFileSync(path.join(__dirname, '../src/banner')).toString());
                console.log(pkg.version);
                console.log('');
                console.log("TypeScript version used by Compodoc : " + Ast.ts.version);
                console.log('');
                if (application.FileEngine.existsSync(cwd + path.sep + 'package.json')) {
                    var packageData = application.FileEngine.getSync(cwd + path.sep + 'package.json');
                    if (packageData) {
                        var parsedData = JSON.parse(packageData);
                        var projectDevDependencies = parsedData.devDependencies;
                        if (projectDevDependencies && projectDevDependencies.typescript) {
                            var tsProjectVersion = application.AngularVersionUtil.cleanVersion(projectDevDependencies.typescript);
                            console.log("TypeScript version of current project : " + tsProjectVersion);
                            console.log('');
                        }
                    }
                }
                console.log("Node.js version : " + process.version);
                console.log('');
                console.log("Operating system : " + osName(os.platform(), os.release()));
                console.log('');
            }
        }
        if (configExplorerResult) {
            if (typeof configExplorerResult.config !== 'undefined') {
                application.logger.info("Using configuration file : " + configExplorerResult.filepath);
            }
        }
        if (!configExplorerResult) {
            application.logger.warn("No configuration file found, switching to CLI flags.");
        }
        if (program.language && !application.I18nEngine.supportLanguage(program.language)) {
            application.logger.warn("The language " + program.language + " is not available, falling back to " + application.I18nEngine.fallbackLanguage);
        }
        if (program.tsconfig && typeof program.tsconfig === 'boolean') {
            application.logger.error("Please provide a tsconfig file.");
            process.exit(1);
        }
        if (configFile.tsconfig) {
            application.Configuration.mainData.tsconfig = configFile.tsconfig;
        }
        if (program.tsconfig) {
            application.Configuration.mainData.tsconfig = program.tsconfig;
        }
        if (program.maxSearchResults) {
            application.Configuration.mainData.maxSearchResults = program.maxSearchResults;
        }
        if (configFile.files) {
            scannedFiles = configFile.files;
        }
        if (configFile.exclude) {
            excludeFiles = configFile.exclude;
        }
        if (configFile.include) {
            includeFiles = configFile.include;
        }
        /**
         * Check --files argument call
         */
        var argv = require('minimist')(process.argv.slice(2));
        if (argv && argv.files) {
            application.Configuration.mainData.hasFilesToCoverage = true;
            if (typeof argv.files === 'string') {
                _super.prototype.setFiles.call(this, [argv.files]);
            }
            else {
                _super.prototype.setFiles.call(this, argv.files);
            }
        }
        if (program.serve && !application.Configuration.mainData.tsconfig && program.output) {
            // if -s & -d, serve it
            if (!application.FileEngine.existsSync(application.Configuration.mainData.output)) {
                application.logger.error(application.Configuration.mainData.output + " folder doesn't exist");
                process.exit(1);
            }
            else {
                application.logger.info("Serving documentation from " + application.Configuration.mainData.output + " at http://" + application.Configuration.mainData.hostname + ":" + program.port);
                _super.prototype.runWebServer.call(this, application.Configuration.mainData.output);
            }
        }
        else if (program.serve && !application.Configuration.mainData.tsconfig && !program.output) {
            // if only -s find ./documentation, if ok serve, else error provide -d
            if (!application.FileEngine.existsSync(application.Configuration.mainData.output)) {
                application.logger.error('Provide output generated folder with -d flag');
                process.exit(1);
            }
            else {
                application.logger.info("Serving documentation from " + application.Configuration.mainData.output + " at http://" + application.Configuration.mainData.hostname + ":" + program.port);
                _super.prototype.runWebServer.call(this, application.Configuration.mainData.output);
            }
        }
        else if (application.Configuration.mainData.hasFilesToCoverage) {
            if (program.coverageMinimumPerFile) {
                application.logger.info('Run documentation coverage test for files');
                _super.prototype.testCoverage.call(this);
            }
            else {
                application.logger.error('Missing coverage configuration');
            }
        }
        else {
            if (program.hideGenerator) {
                application.Configuration.mainData.hideGenerator = true;
            }
            if (application.Configuration.mainData.tsconfig && program.args.length === 0) {
                /**
                 * tsconfig file provided only
                 */
                var testTsConfigPath = application.Configuration.mainData.tsconfig.indexOf(process.cwd());
                if (testTsConfigPath !== -1) {
                    application.Configuration.mainData.tsconfig = application.Configuration.mainData.tsconfig.replace(process.cwd() + path.sep, '');
                }
                if (!application.FileEngine.existsSync(application.Configuration.mainData.tsconfig)) {
                    application.logger.error("\"" + application.Configuration.mainData.tsconfig + "\" file was not found in the current directory");
                    process.exit(1);
                }
                else {
                    var _file = path.join(path.join(process.cwd(), path.dirname(application.Configuration.mainData.tsconfig)), path.basename(application.Configuration.mainData.tsconfig));
                    // use the current directory of tsconfig.json as a working directory
                    cwd = _file
                        .split(path.sep)
                        .slice(0, -1)
                        .join(path.sep);
                    application.logger.info('Using tsconfig file ', _file);
                    var tsConfigFile = application.readConfig(_file);
                    scannedFiles = tsConfigFile.files;
                    if (scannedFiles) {
                        scannedFiles = application.handlePath(scannedFiles, cwd);
                    }
                    if (typeof scannedFiles === 'undefined') {
                        excludeFiles = tsConfigFile.exclude || [];
                        includeFiles = tsConfigFile.include || [];
                        scannedFiles = [];
                        var excludeParser_1 = new ParserUtil(), includeParser_1 = new ParserUtil();
                        excludeParser_1.init(excludeFiles, cwd);
                        includeParser_1.init(includeFiles, cwd);
                        var startCwd = cwd;
                        var excludeParserTestFilesWithCwdDepth = excludeParser_1.testFilesWithCwdDepth();
                        if (!excludeParserTestFilesWithCwdDepth.status) {
                            startCwd = excludeParser_1.updateCwd(cwd, excludeParserTestFilesWithCwdDepth.level);
                        }
                        var includeParserTestFilesWithCwdDepth = includeParser_1.testFilesWithCwdDepth();
                        if (!includeParser_1.testFilesWithCwdDepth().status) {
                            startCwd = includeParser_1.updateCwd(cwd, includeParserTestFilesWithCwdDepth.level);
                        }
                        var finder = require('findit2')(startCwd || '.');
                        finder.on('directory', function (dir, stat, stop) {
                            var base = path.basename(dir);
                            if (base === '.git' || base === 'node_modules') {
                                stop();
                            }
                        });
                        finder.on('file', function (file, stat) {
                            if (/(spec|\.d)\.ts/.test(file)) {
                                application.logger.warn('Ignoring', file);
                            }
                            else if (excludeParser_1.testFile(file) &&
                                path.extname(file) === '.ts') {
                                application.logger.warn('Excluding', file);
                            }
                            else if (includeFiles.length > 0) {
                                /**
                                 * If include provided in tsconfig, use only this source,
                                 * and not files found with global findit scan in working directory
                                 */
                                if (path.extname(file) === '.ts' && includeParser_1.testFile(file)) {
                                    application.logger.debug('Including', file);
                                    scannedFiles.push(file);
                                }
                                else {
                                    if (path.extname(file) === '.ts') {
                                        application.logger.warn('Excluding', file);
                                    }
                                }
                            }
                            else {
                                application.logger.debug('Including', file);
                                scannedFiles.push(file);
                            }
                        });
                        finder.on('end', function () {
                            _super.prototype.setFiles.call(_this, scannedFiles);
                            if (program.coverageTest || program.coverageTestPerFile) {
                                application.logger.info('Run documentation coverage test');
                                _super.prototype.testCoverage.call(_this);
                            }
                            else {
                                _super.prototype.generate.call(_this);
                            }
                        });
                    }
                    else {
                        _super.prototype.setFiles.call(this, scannedFiles);
                        if (program.coverageTest || program.coverageTestPerFile) {
                            application.logger.info('Run documentation coverage test');
                            _super.prototype.testCoverage.call(this);
                        }
                        else {
                            _super.prototype.generate.call(this);
                        }
                    }
                }
            }
            else if (application.Configuration.mainData.tsconfig && program.args.length > 0) {
                /**
                 * tsconfig file provided with source folder in arg
                 */
                var testTsConfigPath = application.Configuration.mainData.tsconfig.indexOf(process.cwd());
                if (testTsConfigPath !== -1) {
                    application.Configuration.mainData.tsconfig = application.Configuration.mainData.tsconfig.replace(process.cwd() + path.sep, '');
                }
                var sourceFolder = program.args[0];
                if (!application.FileEngine.existsSync(sourceFolder)) {
                    application.logger.error("Provided source folder " + sourceFolder + " was not found in the current directory");
                    process.exit(1);
                }
                else {
                    application.logger.info('Using provided source folder');
                    if (!application.FileEngine.existsSync(application.Configuration.mainData.tsconfig)) {
                        application.logger.error("\"" + application.Configuration.mainData.tsconfig + "\" file was not found in the current directory");
                        process.exit(1);
                    }
                    else {
                        var _file = path.join(path.join(process.cwd(), path.dirname(application.Configuration.mainData.tsconfig)), path.basename(application.Configuration.mainData.tsconfig));
                        // use the current directory of tsconfig.json as a working directory
                        cwd = _file
                            .split(path.sep)
                            .slice(0, -1)
                            .join(path.sep);
                        application.logger.info('Using tsconfig file ', _file);
                        var tsConfigFile = application.readConfig(_file);
                        scannedFiles = tsConfigFile.files;
                        if (scannedFiles) {
                            scannedFiles = application.handlePath(scannedFiles, cwd);
                        }
                        if (typeof scannedFiles === 'undefined') {
                            excludeFiles = tsConfigFile.exclude || [];
                            includeFiles = tsConfigFile.include || [];
                            scannedFiles = [];
                            var excludeParser_2 = new ParserUtil(), includeParser_2 = new ParserUtil();
                            excludeParser_2.init(excludeFiles, cwd);
                            includeParser_2.init(includeFiles, cwd);
                            var startCwd = sourceFolder;
                            var excludeParserTestFilesWithCwdDepth = excludeParser_2.testFilesWithCwdDepth();
                            if (!excludeParserTestFilesWithCwdDepth.status) {
                                startCwd = excludeParser_2.updateCwd(cwd, excludeParserTestFilesWithCwdDepth.level);
                            }
                            var includeParserTestFilesWithCwdDepth = includeParser_2.testFilesWithCwdDepth();
                            if (!includeParser_2.testFilesWithCwdDepth().status) {
                                startCwd = includeParser_2.updateCwd(cwd, includeParserTestFilesWithCwdDepth.level);
                            }
                            var finder = require('findit2')(path.resolve(startCwd));
                            finder.on('directory', function (dir, stat, stop) {
                                var base = path.basename(dir);
                                if (base === '.git' || base === 'node_modules') {
                                    stop();
                                }
                            });
                            finder.on('file', function (file, stat) {
                                if (/(spec|\.d)\.ts/.test(file)) {
                                    application.logger.warn('Ignoring', file);
                                }
                                else if (excludeParser_2.testFile(file)) {
                                    application.logger.warn('Excluding', file);
                                }
                                else if (includeFiles.length > 0) {
                                    /**
                                     * If include provided in tsconfig, use only this source,
                                     * and not files found with global findit scan in working directory
                                     */
                                    if (path.extname(file) === '.ts' &&
                                        includeParser_2.testFile(file)) {
                                        application.logger.debug('Including', file);
                                        scannedFiles.push(file);
                                    }
                                    else {
                                        if (path.extname(file) === '.ts') {
                                            application.logger.warn('Excluding', file);
                                        }
                                    }
                                }
                                else {
                                    application.logger.debug('Including', file);
                                    scannedFiles.push(file);
                                }
                            });
                            finder.on('end', function () {
                                _super.prototype.setFiles.call(_this, scannedFiles);
                                if (program.coverageTest || program.coverageTestPerFile) {
                                    application.logger.info('Run documentation coverage test');
                                    _super.prototype.testCoverage.call(_this);
                                }
                                else {
                                    _super.prototype.generate.call(_this);
                                }
                            });
                        }
                        else {
                            _super.prototype.setFiles.call(this, scannedFiles);
                            if (program.coverageTest || program.coverageTestPerFile) {
                                application.logger.info('Run documentation coverage test');
                                _super.prototype.testCoverage.call(this);
                            }
                            else {
                                _super.prototype.generate.call(this);
                            }
                        }
                    }
                }
            }
            else {
                application.logger.error('tsconfig.json file was not found, please use -p flag');
                outputHelp();
            }
        }
    };
    return CliApplication;
}(application.Application));

exports.CliApplication = CliApplication;
